### ISSUE TITLE
[BUG] Minimax m2.7 mxfp4 thinks in chinese

### ISSUE THREAD

**Issue by leanzero-srl** [exo-explore/exo#1924]
## Describe the bug

Minimax m2.7 mxfp4 thinks in chinese

## To Reproduce

1. Download minimax m2.7 mxfp4
2. Run with qwen code
3.Thinking is in chinese

## Expected behavior

English...
## Actual behavior

Chinese
## Environment

- macOS Version:
- EXO Version: 1.0.70
- Hardware:
  - Device 1: M4 max 128gb
  - Device 2: M3 ultra 96gb
  - Additional devices:
- Interconnection:
  - (e.g., Thunderbolt 4 cable between Device 1 and 2)
  - (e.g., WiFi 6 for Device 3)
  - (e.g., 10GbE Ethernet between all devices)

**Comment by leanzero-srl** [comment]
The 4 bit one does the same thing.

**Comment by leanzero-srl** [comment]
Tried with claude and it's the same thing. It';s not only that it answers in chinese, it also endlessly loops.

[exo.log](https://github.com/user-attachments/files/26854437/exo.log)

**Comment by leanzero-srl** [comment]
This resolves it to an extent where it actually stops chinese thinking and gibberish 

1) Repetition loops on MiniMax-M2.x

**Symptom.** MiniMax-M2.x running under exo produced endless repetition loops (often in Chinese) on typical chat-completion requests. Reproduced on both M2.5 and M2.7 at 4-bit; indifferent to prompt content.

**Root cause.** `TextGenerationTaskParams.repetition_penalty` defaults to `None`. mlx_lm's `make_logits_processors` treats `None` as "install no repetition-penalty processor at all":
```python
# mlx_lm/sample_utils.py line ~117-124
for make_penalty, penalty, context_size in repetition_penalties:
    if penalty is not None and penalty != 0:
        logits_processors.append(make_penalty(penalty, context_size))
```
Clients like Claude Code / Qwen Code don't set `repetition_penalty` in the request body, so it arrives `None` and the model generates with zero suppression on repeating tokens. Thinking models degrade into loops fast under that config.

**Fix.** Per-model sampling defaults in the model card, resolved at the runner boundary.

- `src/exo/shared/models/model_cards.py` — new optional `SamplingDefaults` nested model on `ModelCard`. Fields: `temperature`, `top_p`, `top_k`, `min_p`, `repetition_penalty`, `repetition_context_size`, all optional.
- `src/exo/worker/engines/mlx/utils_mlx.py` — new `apply_sampling_defaults_from_card(task_params, card)`. Fills in fields the request left as `None` from the card's `sampling_defaults`. Request-supplied values always win over card defaults.
- `src/exo/worker/runner/llm_inference/runner.py` — `submit_text_generation` calls the resolver before handing the task to the generator.

Also added a defensive coalesce in `generate.py` and `batch_generate.py`: `make_repetition_penalty(penalty, context_size)` crashes on `tokens[-None:]` when `context_size=None`, so we clamp None → 20 (mlx_lm's own library default) right before `make_logits_processors`.

**Card populations (upstream-verified values).** MiniMax's own model cards (`MiniMaxAI/MiniMax-M2.5`, `MiniMaxAI/MiniMax-M2.7`) recommend `temperature=1.0, top_p=0.95, top_k=40`. They don't prescribe a `repetition_penalty`, but a penalty is needed to kill the loops, so `1.1` added per in-cluster reproducer feedback.

Populated:
- `resources/inference_model_cards/mlx-community--MiniMax-M2.5-{4bit,6bit,8bit}.toml`
- `resources/inference_model_cards/mlx-community--MiniMax-M2.7-{bf16,4bit,4bit-mxfp4,5bit,6bit,8bit}.toml`

Any card without a `[sampling_defaults]` block is unaffected — the resolver is a no-op.

### 2) Failed tool calls

**Symptoms.** Three distinct failure modes reaching the client:
- **(a)** Casing drift: schema declares `file_path`, model emits `filePath`. The canonical parser returned a valid tool-call object but the downstream client rejected it with "unexpected parameter filePath / missing required parameter file_path".
- **(b)** Opening-delimiter-only hallucination: model emits the documented `<minimax:tool_call>` (or `<tool_call>`) open token but never emits the matching close, tailing off into prose or an Anthropic-style `<Read filePath="..."/>` self-closing tag instead. `parse_tool_calls` silently buffered every subsequent token waiting for a close that never arrived — the stream appeared to "stop" mid-response.
- **(c)** Whole-response wrong format: canonical open and close delimiters present, but the content between them was self-closing tags or Qwen-Code-style `<tool name="X" ...>` attribute lists. Inner parser returned `None`, entire buffered text was emitted with `finish_reason=error`.

**Fixes.** Three layered changes in `src/exo/worker/runner/llm_inference/`:

1. **Key-case normalization** in `tool_parsers.py`. New `_case_variants(name)` returns plausible camelCase ↔ snake_case drifts. New `_remap_keys_to_schema(emitted, schema_properties)` rewrites emitted arg keys to match the schema's canonical names when case-variants match. Hooked into `_coerce_tool_calls_to_schema` before the existing type coercion. Exact-match keys are untouched; unknown keys pass through unchanged. Addresses (a).

2. **Permissive fallback parser** in `tool_parsers.py`. New `fallback_parse_hallucinated_tool_calls(text, tools)` uses regex to extract self-closing tags (`<ToolName arg="val"/>`) and generic attribute-list wrappers (`<tool name="ToolName" arg="val"/>`). Allow-lists against the `tools` kwarg to avoid misreading regular XML in prose as tool calls. Reuses the key-remap + schema coercion internally. Addresses (c).

3. **Buffer escape valve** in `model_output_parsers.py` `parse_tool_calls`. Hard 8 KiB cap on the tool-call buffer; when exceeded without a close delimiter, runs the fallback parser on what's buffered, rescues if possible, otherwise flushes the buffered text as a plain `GenerationResponse` (no `finish_reason`) so the stream keeps moving. The `finish_reason` end-of-stream path also gets one last fallback attempt before emitting the "error" response. Addresses (b); the user-visible "inference stopped" symptom goes away even when the model keeps emitting garbage.


HOWEVER the tool calls are still completely busted.

**Comment by qtkt2hmr2v-debug** [comment]
## Additional failure mode: `enable_thinking=false` → empty response bodies (60% failure rate)

Adding quantified data that seems related to the same root cause described above.

**Setup:** Apple Mac Studio M4 Max, 512GB unified RAM, EXO v1.0.70, single-node MLX.

**Observation:** When calling `mlx-community/MiniMax-M2.7-6bit` with `enable_thinking=false`, 60% of requests return HTTP 200 but with unusable content — either an empty body, `choices: []`, or `content: ""`.

| Mode | Failures | Notes |
|------|----------|-------|
| `enable_thinking=false` | 8/20 HTTP errors + 4/20 empty `content` = **12/20 unusable** | Strongly correlated with prompt complexity |
| `enable_thinking=true`  | 3/20 failures | Partially usable |
| MiniMax-M2.5-6bit (same HW) | 0–1/20 | Stable in both modes |

**Failure pattern by prompt complexity:**
- Simple prompts (1–2 sentences): mostly succeed
- Medium complexity prompts (code generation, explanations): ~80% failure rate
- Complex reasoning prompts (architecture, algorithms): 4/5 failures

**Hypothesis:** This may be related to the `repetition_penalty=None` issue you identified above. Without `repetition_penalty`, the model likely enters a degenerate state early in generation when thinking is suppressed — producing tokens internally but emitting nothing to the output stream. The commit note for PR #1884 also states: *"the chat template always emits `<think>` tags (no toggle)"* — so `enable_thinking=false` may be fundamentally unsupported for M2.7 at the model level, not just a sampling issue.

**Note:** We also observed that even with `enable_thinking=false`, M2.5 still generates non-zero `reasoning_tokens` in the API response (115–409 rtok depending on prompt complexity), confirming the model always thinks internally. The `enable_thinking` flag only controls whether `<think>` tags appear in the output — not whether the model reasons.

The `repetition_penalty=1.1` fix from your analysis likely addresses the loop/Chinese thinking issue. It may also reduce the empty-response rate for `enable_thinking=false`, but we haven't been able to test this yet since the fix isn't in a released version.

Happy to share the raw benchmark JSON if useful.

**Comment by leanzero-srl** [comment]
Switching to pipeline instead of tensor does a lot of wonder to the tool calls. You could even say it's almost working really well. The problem with pipeline is obviously that it's quite a bit slower :(

**Comment by rltakashige** [comment]
@leanzero-srl We'll be making a release today with this fix! Going to close this as completed, please let me know if you encounter any other issues - thanks!

**Comment by leanzero-srl** [comment]
@rltakashige  But wait, so is this tested with Tensor and RDMA? Have you actually fixed tool calling? What kind of testing have you guys done here?
